Design Document: Omega-0 algorithm formulation (#33)
E3SM-Project/Omega: components/omega/doc/design/OmegaV0ShallowWater.md

Like the previous requirement, tests will first be conducted with layered shallow water equations and later with additional tracer advection. For reference, [Bishnu et al. 2023](https://gmd.copernicus.org/articles/16/5539/2023) was able to obtain nearly identical throughput between 64 CPU cores and a single GPU on Perlmutter using a Julia code and a layered shallow water test case.

## 3 Algorithmic Formulation

### 3.1 Governing Equations

#### 3.1.1 Continuous Equations

The algorithms for Omega-0 are given here in full detail. The original version with derivations are given in [Thuburn et al. 2009](https://www.sciencedirect.com/science/article/pii/S0021999109004434) and [Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780).
In continuous form, these are

$$
\frac{\partial \boldsymbol{u}}{\partial t} + \left(\boldsymbol{u}\cdot\nabla\right)\boldsymbol{u}  + f \boldsymbol{k}\times\boldsymbol{u}

> = -g\nabla(h+b)

Reviewer: Shouldn't `(h+b)` be `(h+H)` according to the definitions of `H` in the table below? Or need to change `H` in the table to `b` as bottom topography?

Author: Done. I changed H to b.

$$

### 3.2 Variable Definitions

Table 1. Definition of variables

| symbol  | name             | units    | location | name in code | notes  |
|---------------------|-----------------------------|----------|-|---------|-------------------------------------------------------|
| $C_D$               | bottom drag                 | 1/m      | constant |   |                                                              |
| $C_W$               | wind stress coefficient     | 1/m      | constant |   |                                                              |
| $\mathcal{D} $      | drag                        | m/s^2    | edge     |   |                                                              |
| $f$                 | Coriolis parameter          | 1/s      | vertex   |   |                                                              |
| $\mathcal{F} $      | forcing                     | m/s^2    | edge     |   |                                                              |
| $g$                 | gravitational acceleration  | m/s^2    | constant |   |                                                              |

> | $H$                 | total unperturbed depth     | m        | cell     |   |                                                              |

Reviewer: This suggestion is resolved by @mark-petersen .

[Thuburn et al. 2009](https://www.sciencedirect.com/science/article/pii/S0021999109004434) and [Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780), as described in the algorithmic formulation in Section 3 below. This is the same base formulation as in MPAS-Ocean. In addition, we will consider small alterations from the original MPAS-Ocean horizontal discretization and include them as options if they are of minimal additional effort. This includes the recent AUST formulation in [Calandrini et al. 2021](https://www.sciencedirect.com/science/article/pii/S146350032100161X) and simple vorticity averaging considered by the Omega team this past year.

### 2.3  Omega-0 will use MPAS format unstructured-mesh domains

We will continue to use the MPAS-format netcdf files for input and output, with the same mesh variable names and dimensions. This will facilitate ease of use and interoperability between MPAS-Ocean and Omega.

Omega mesh information will be stored in a separate mesh file, and not be included with initial condition, restart, or output files. This way there is never redundant mesh data stored in these files.

> To do: In the design section below, add list of minimum number of variables that defines the mesh, and the additional variables that can be computed on start-up.

Reviewer 2: Do we want to still do this?

Author: Good point. I think that is not needed here.

Like the previous requirement, tests will first be conducted with layered shallow water equations and later with additional tracer advection. For reference, [Bishnu et al. 2023](https://gmd.copernicus.org/articles/16/5539/2023) was able to obtain nearly identical throughput between 64 CPU cores and a single GPU on Perlmutter using a Julia code and a layered shallow water test case.

## 3 Algorithmic Formulation

### 3.1 Governing Equations

#### 3.1.1 Continuous Equations

The algorithms for Omega-0 are given here in full detail, with all variables defined in section 3.2 below.
We begin with the standard shallow water equations, which are derived from conservation of momentum and mass in a single layer with uniform density in a rotating frame. The standard presentation may be found in standard textbooks on Geophysical Fluid Dynamics, such as those by Vallis (2017), Cushman‐Roisin and Beckers (2011), Pedlosky (1987), and Gill (2016). In continuous form, these are

$$
\frac{\partial \boldsymbol{u}}{\partial t} + \left(\boldsymbol{u}\cdot\nabla\right)\boldsymbol{u}  + f \boldsymbol{k}\times\boldsymbol{u}

> = -g\nabla(h+b)

Reviewer 2: Do we want $b$ to follow the traditional positive-down convention since that's what we use in the code? It's also fine to leave as-is since this is consistent with Ringler 2010.

Author: Good point. I changed to `-b` so it matches the `BottomDepth` variable.

|---------------------|-----------------------------|----------|-|---------|-------------------------------------------------------|
| $b$                 | height of bottom (pos. up)  | m        | cell     |   |                                                              |
| $C_D$               | bottom drag                 | 1/m      | constant |   |                                                              |
| $C_W$               | wind stress coefficient     | 1/m      | constant |   |                                                              |
| $\mathcal{D} $      | drag                        | m/s^2    | edge     |   |                                                              |
| $f$                 | Coriolis parameter          | 1/s      | vertex   | fVertex  |                                                              |
| $\mathcal{F} $      | forcing                     | m/s^2    | edge     |   |                                                              |
| $g$                 | gravitational acceleration  | m/s^2    | constant |   |                                                              |

> | $h$                 | thickness of layer          | m        | cell     | layerThickness  |                                                              |

Reviewer 2: OMEGA variable names start with a capital letter

Author: Thanks. I made that change. The variable names really go in the design section below, but it is so convenient to put here that I started to fill it in.

| $x_v$ | point | Location of center of dual-mesh cells (Vertex)| (xVertex, yVertex, zVertex) | (lonVertex, latVertex) |
| $d_e$ | line | segment distance between neighboring $x_i$ locations | dcEdge | |
| $l_e$ | line | segment distance between neighboring $x_v$ locations | dvEdge | |
| $P_i$ | cell | A cell on the primal mesh | | |
| $D_v$ | cell | A cell on the dual-mesh | | |

![Variable positions for MPAS mesh specification.](../_static/variablePosition.png "Variable positions")

Figure 1. Variable positions for MPAS mesh specification.

Table 3. Definition of element groups used to build the discrete system.

| Syntax              | Definition                  | MPAS mesh name | 
|---------------------|-----------------------------|---|
| $e\in EC(i)$ |  Set of edges that define the boundary of $P_i$ | edgesOnCell |
| $e\in EV(v)$ |  Set of edges that define the boundary of $D_v$ | edgesOnVertex |
| $i\in CE(e)$ |  Two primal mesh cells that share edge $e$ | cellsOnEdge |
| $i\in CV(v)$ |  Set of primal mesh cells that form the vertices of dual mesh cell $D_v$ | cellsOnVertex |
| $v\in VE(e)$ |  The two dual-mesh cells that share edge $e$ | verticesOnEdge |

> | $v\in VI(i)$ |  The set of dual-mesh cells that form the vertices of primal mesh cell $P_i$ | verticesOnCell |

Author: Using the notation in [Ringler 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780#tbl2), the `i` is used for cell center. I think they chose that because subscript `c` and `e` look the same.

> | $v\in VI(i)$ |  The set of dual-mesh cells that form the vertices of primal mesh cell $P_i$ | verticesOnCell |

Reviewer 3: @mark-petersen I am not insisting on this change, but I think that this is simply a typo in the table in Ringler 2010. See Eq (26) therein where this variable is called `VC` and Table 2.2 in a later [TRiSK extension paper](https://doi.org/10.1137/110850293).

> | $v\in VI(i)$ |  The set of dual-mesh cells that form the vertices of primal mesh cell $P_i$ | verticesOnCell |

Author: Thanks. You are right, I see that now. It was also correct in the original [Thuburn et al. 2009](http://www.sciencedirect.com/science/article/pii/S0021999109004434).

#### 3.2.4. Perpendicular vector component

The native component for the prognostic velocity field is normal edge velocity $u_e$. The perpendicular component at the edge is computed diagnostically as
([Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780) eqn 24)

$$
F_e^\perp = \frac{1}{d_e} \sum_{e'\in ECP(e)} w_{e,e'} \, l_e \, F_{e'}.
$$

The perpendicular vector is computed from the edge-normal components of all the edges of the two neighboring cells of edge $e$, denoted as $e'\in ECP(e)$. For example, on a mesh with hexagons, there are a total of 10 edges $e'$ on the two cells neighboring edge $e$ (the original edge $e$ is not included).  The perpendicular velocity $u_e^\perp$ is used in the Coriolis force, which appears in the potential vorticity advection term.  The weighting coefficients $w_{e,e'}$ are carefully chosed to conserve potential vorticity, as described in
[Thuburn et al. 2009](https://www.sciencedirect.com/science/article/pii/S0021999109004434)
and
[Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780).

#### 3.2.5. Perpendicular Gradient

> This maps a vertex scalar to an edge-tangential vector component (check $t_{e,v}$)

Author: Thanks. I updated the document to explain this, and added the variable to the table.

See [Calandrini et al. 2021](https://www.sciencedirect.com/science/article/pii/S146350032100161X) section 2.3 for more information.

#### 3.3.2. Potential vorticity term

The potential vorticity term, $q\left(h\boldsymbol{u}^{\perp}\right)$, includes the rotational part of the advection. It may be computed in two ways,

$$
\text{Option 1:  } \left[ \frac{\omega_v +f_v}{[h_i]_v}\right]_e\left([h_i]_e u_e^{\perp}\right)
$$

$$
\text{Option 2:  }([\omega_v]_e +f_e) u_e^\perp
$$

The first computes the potential vorticity $q_v$ at the vertex and interpolates that quantity to the edge, which is what is done in MPAS-Ocean. One may also cancel the thickness $h$ (ignoring the interpolated locations) and use option 2. Additional interpolation options and results are presented in [Calandrini et al. 2021](https://www.sciencedirect.com/science/article/pii/S146350032100161X)

#### 3.3.3. Sea surface height gradient
The sea surface height (SSH) gradient uses the standard gradient formulation from cell center to edge,

$$
-g\nabla(h_i+b_i) = -g\frac{1}{d_e} \sum_{i\in CE(e)} -n_{e,i}(h_i+b_i).
$$

#### 3.3.4. Del2 momentum dissipation

> The Del2, or Laplacian operator, viscous momentum dissipation maps edge-normal velocity back to the edge-normal component of the Laplacian. In TRiSK this is done with the "grad-div plus vorticity curl" formulation. (check grad omega term).

Author: Thanks. I updated the document to explain this

$$

> \nu_2 \nabla^2 u_e = \nu_2 \left( \nabla \delta_i + {\bf k} \times \nabla \omega_v\right) = \frac{\nu_2}{d_e} \sum_{i\in CE(e)} -n_{e,i}\delta_i + \frac{\nu_2}{l_e} \sum_{v\in VE(e)} -n_{e,v}\omega_v

Author: Thanks. That is divergence. I will add it.

$$
- Ra \: u_e
$$

#### 3.3.7. Bottom drag
Bottom drag is more relevant to layered models than to shallow water systems, but is included here for completeness. It is a quadratic drag applied to the edge velocity,

$$
- C_D \frac{u_e\left|u_e\right|}{[h_i]_e}.
$$

#### 3.3.8. Wind forcing
Wind forcing has the same form as the bottom drag, but the forcing is the difference between the current velocity and the wind $u_W$, interpolated and projected to the edge normal direction.

$$
- C_W \frac{(u_W - u_e)\left|u_W - u_e\right|}{[h_i]_e}
$$

### 3.4 Thickness and Tracer Terms

#### 3.4.1. Tracer advection

There are many schemes available for tracer advection. Simple schemes include centered advection and upstream. MPAS-Ocean uses Flux Corrected Transport, which is fourth order under normal conditions, and reduces to third order to preserve monotinicity. 

The tracer advection term is
$$
\nabla \cdot \left(u_e [h_i \phi_i]_e \right)
$$

and the stencil for divergence makes this
$$
\nabla \cdot \left( u_e [h_i \phi_i]_e \right)  =  \frac{1}{A_i} \sum_{e\in EC(i)} n_{e,i}u_e [h_i \phi_i]_e l_e
$$

> The question is how to interpolate to the edge to obtain $[h_i \phi_i]_e$. Centered advection simply uses the midpoint rule. Upwind takes the value from the upstream cell. Higher order schemes cast a wider stencil to compute the value at this edge.

Reviewer 2: We should add the higher order scheme here down the road.

Author: Agreed. I added a sentence that MPAS-Ocean used FCT.

The divergence operator above is applied to a general vector field $\bf F$. 
In the actual formulations below, we substitute the velocity at the edge $F_e = u_e$ for the divergence variable in the momentum equation, and the thickness-weighted tracer
$F_e = u_e [h_i \phi_i]_e$
 in the tracer advection term. To be clear, when we refer to the divergence variable, rather than the operator, this means
$$
\left( \nabla \cdot {\bf u}\right)_i =
D_i
=  \frac{1}{A_i} \sum_{e\in EC(i)} n_{e,i} \, u_e \, l_e.
$$

#### 3.2.2. Gradient

The gradient operator maps a cell-centered scalar to an edge-normal vector component
([Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780) eqn 22) is

$$
\left( \nabla h \right)_e
\equiv
\nabla h_i = \frac{1}{d_e} \sum_{i\in CE(e)} -n_{e,i} \, h_i,
$$
where $i\in CE(e)$ identifies the two cells that neighbor an edge. The edge subscript is dropped, so $\nabla h_i$ is understood to occur at an edge.

For practical purposes the sign indicator variable $n_{e,i}$ is not used in the code, and we opt for the simpler formulation
$$
\left( \nabla h \right)_e
\equiv
\nabla h_i = \frac{h_{i2} - h_{i1} }{d_e},
$$

> where $\left\{i_1, i_2\right\}$ are the cells neighboring edge $e$.  The indices $\left\{i_1, i_2\right\}$ are ordered such that the normal vector ${\bf n}$ points from cell $i_1$ to cell $i_2$. In the code ${\bf n}$ points from CellsOnEdge(0,iEdge) to CellsOnEdge(1,iEdge).

Reviewer 3: ```suggestion
where $\left\{i_1, i_2\right\}$ are the cells neighboring edge $e$.  The indices $\left\{i_1, i_2\right\}$ are ordered such that the normal vector ${\bf n}$ points from cell $i_1$ to cell $i_2$. In the code ${\bf n}$ points from CellsOnEdge(IEdge, 0) to CellsOnEdge(IEdge, 1).
```

${\bf x}_v$, then $t_{e,v}=1$. The summation is over $e\in EV(v)$, which are the edges that terminate at vertex $v$. There are *always* three edges that terminate at each vertex for Voronoi Tessellations, and four edges on each vertex for quadrilateral meshes, unless neighboring cells are missing for land boundaries. Again, the subscript $v$ is dropped and a vertex is assumed for the location of the curl.

#### 3.2.4. Perpendicular vector component

The perpendicular vector component a vector field is defined as
$$
 \nabla^\perp {\bf F} \equiv
 {\bf k} \times \nabla {\bf F}
$$
so that it points 90$^o$ to the left of ${\bf F}$ in a right-handed coordinate system. 

In TRiSK, the native component for the prognostic velocity field is normal edge velocity $u_e$. The perpendicular component at the edge is computed diagnostically as
([Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780) eqn 24)

$$

> u_e^\perp = \frac{1}{d_e} \sum_{e'\in ECP(e)} w_{e,e'} \, l_e \, u_{e'}.

Reviewer 3: The index on `l` should be `e'`. There is also another issue. While this agrees with the TRiSK paper, it does not agree with Omega code. I believe that in MPAS-Ocean mesh files the reconstruction weights are pre-multiplied by `l_e' / d_e`. It would be good to make a note about this here. The weights also need to be added to the variables table.

Author: Thanks for your detailed review! I fixed this and added text to explain what is used in the code. See the [revised section in the compiled version here](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-omega0-sw/html/design/OmegaV0ShallowWater.html#perpendicular-vector-component).

[Thuburn et al. 2009](https://www.sciencedirect.com/science/article/pii/S0021999109004434)
and
[Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780).

#### 3.2.5. Perpendicular Gradient
The gradient of a scalar at the middle of an edge, pointing tangentially along the edge (from one vertex to the other), is sometimes used. For example, the del2 formulation requires the perpendicular gradient of vorticity. This is called the perpendicular gradient because the standard gradient is normal to the edge.

The perpendicular gradient maps a scalar at vertices to an edge-tangential vector component,
$$
\left(\nabla \omega_v\right)_e^\perp = \frac{1}{l_e} \sum_{v\in VE(e)} -t_{e,v}\omega_v.
$$
Like the standard gradient, in practice we drop the sign indicator $t_{e,v}$ and rewrite it using the index ordering,
$$
\left(\nabla \omega_v\right)_e^\perp = \frac{\omega_{v2} - \omega_{v1}}{l_e},
$$

> where the positive vector ${\bf n}^\perp$ is 90$^o$ to the left of ${\bf n}$. The indices are ordered such that ${\bf n}^\perp$ points from $v_1$ to $v_2$, which corresponds to VerticesOnEdge(0,iEdge) and VerticesOnEdge(1,iEdge) in the code.

Reviewer 3: ```suggestion
where the positive vector ${\bf n}^\perp$ is 90$^o$ to the left of ${\bf n}$. The indices are ordered such that ${\bf n}^\perp$ points from $v_1$ to $v_2$, which corresponds to VerticesOnEdge(IEdge, 0) and VerticesOnEdge(IEdge, 1) in the code.
```

#### 3.2.6. Cell to Edge Interpolation
This is simply a mid-point averaging from cell centers to the adjoining edge.

$$
[h_i]_e = \frac{1}{2} \sum_{i\in CE(e)} h_i
$$

#### 3.2.7. Vertex to Edge Interpolation
This is a mid-point averaging from vertices to the middle of the connecting edge.

$$
[\omega_v]_e = \frac{1}{2} \sum_{v\in VE(e)} \omega_v
$$

#### 3.2.8. Cell to Vertex Interpolation

> This is an area-weighted average at a vertex from the three surrounding cells. Here ${\tilde A}_v$ is the dual triangle area centered on the vertex, and ${\hat A}_{i,v}$ is the kite area of cell $i$ with the kite point at vertex $v$.

Reviewer 3: `A_v` is defined without the tilde in the curl equation and is missing from the variables table.

Author: Done, thanks.

The prognostic velocity variable on the edge is the edge normal velocity, $u_e$. The tangential velocity $u_e^\perp$ is computed diagnostically from $u_e$. In addition, the full vector may be computed at the cell center from the edge normal velocities $u_e$ of the edges on that cell. That is done in MPAS with radial basis functions, and is explained in this [previous design document](https://github.com/MPAS-Dev/MPAS-Documents/blob/master/shared/rbf_design/rbf.pdf).

### 3.3 Momentum Terms

The computation of each term in (4-6) is now described in detail, along with alternative formulations.

#### 3.3.1. Kinetic energy gradient

The kinetic energy gradient term is the non-rotational part of the nonlinear advection. Fundamentally, it maps the prognostic edge-normal velocity from edges back to an edge scalar quantity. We use the standard gradient formulation from cell center to edge,

$$
\nabla K_i = \frac{1}{d_e} \sum_{i\in CE(e)} -n_{e,i}K_i.
$$

There are several methods to compute the cell-centered kinetic energy. The first uses the edge-normal velocities as in
[Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780) equation 63,

$$

> K_i = \frac{1}{A_i} \sum_{e\in EC(c)} \frac{{\bar A}_{e}}{4}u_e^2

Reviewer 3: `A_e` is not defined and is missing from the variables table.

Author: done, thanks.

## 2 Requirements

### 2.1  Omega-0 will solve the nonlinear shallow water equations, plus inactive tracers

The governing equations for Omega-0 are the shallow water equations, as described in [Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780) eqns 2-7, plus the tracer advection-diffusion equation. These equations are derived from conservation of momentum, volume, and tracers in a single layer. The code will solve a multi-layer formulation with independent, redundant layers in order to test performance with a vertical array dimension. The exact formulation is specified section 3.1 below, with variable definitions in section 3.2.

Boundary conditions will include both no-slip and free-slip. The original MPAS-Ocean implementation had no-slip boundaries, implemented by setting vorticity to zero at the boundaries. The free-slip implementation is described by Darren Engwirda in a [github discussion](https://github.com/E3SM-Ocean-Discussion/E3SM/pull/49).

### 2.2  Numerical method will be the TRiSK formulation

The horizontal discretization will be taken from

> [Thuburn et al. 2009](https://www.sciencedirect.com/science/article/pii/S0021999109004434) and [Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780), as described in the algorithmic formulation in Section 3 below. This is the same base formulation as in MPAS-Ocean. In addition, we will consider small alterations from the original MPAS-Ocean horizontal discretization and include them as options if they are of minimal additional effort. This includes the recent AUST formulation in [Calandrini et al. 2021](https://www.sciencedirect.com/science/article/pii/S146350032100161X) and simple vorticity averaging considered by the Omega team this past year.

Reviewer 4: suggest a minor alteration
```
include them as options if they are of minimal additional effort
```
to 
```
include them as options if they are of minimal additional effort and improve the simulated climate.
```

Author: added, thanks.

Author: added, thanks.

### 2.2  Numerical method will be the TRiSK formulation

The horizontal discretization will be taken from
[Thuburn et al. 2009](https://www.sciencedirect.com/science/article/pii/S0021999109004434) and [Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780), as described in the algorithmic formulation in Section 3 below. This is the same base formulation as in MPAS-Ocean. In addition, we will consider small alterations from the original MPAS-Ocean horizontal discretization and include them as options if they are of minimal additional effort and improve the simulated climate. This includes the recent AUST formulation in [Calandrini et al. 2021](https://www.sciencedirect.com/science/article/pii/S146350032100161X) and simple vorticity averaging considered by the Omega team this past year.

### 2.3  Omega-0 will use MPAS format unstructured-mesh domains

> We will continue to use the MPAS-format netcdf files for input and output, with the same mesh variable names and dimensions. This will facilitate ease of use and interoperability between MPAS-Ocean and Omega.

Reviewer 4: what is mean by 'MPAS-format netcdf file'?  @xylar are MPAS outputs CF compliant now?  I also assume we won't use the same variable names right?

Author: This means that we continue to use the same variables in the MPAS mesh specifications, so that MPAS meshes may be used for Omega. 

We are planning to use the same variable names for mesh variables, except that we are changing from camelCase to PascalCase (first letter caps for consistency). We can still read the same files with this case change.

Reviewer 4: Got it, thanks for the explanation.  That sounds good as you describe.

&= \omega \boldsymbol{u}^{\perp} + \nabla K.
$$

The governing equations for Omega-0 in continuous form are then
$$
\frac{\partial \boldsymbol{u}}{\partial t} + q\left(h\boldsymbol{u}^{\perp}\right) = -g\nabla(h-b) - \nabla K + \nu_2 \nabla^2 \boldsymbol{u} - \nu_4 \nabla^4 \boldsymbol{u} + \mathcal{D} + \mathcal{F}
\hspace{1cm}   (1)
$$

$$
\frac{\partial h}{\partial t} + \nabla \cdot \left(h \boldsymbol{u}\right) = 0,
\hspace{1cm}   (2)
$$

$$
\frac{\partial h \phi}{\partial t} + \nabla \cdot \left(h \boldsymbol{u} \phi\right) = \kappa_2 h \nabla^2 \phi - \kappa_4 h \nabla^4 \phi.
\hspace{1cm}   (3)
$$
These are typically referred to as the momentum equation (or velocity equation), the thickness equation, and the tracer equation. 

The additional momentum terms in (1) are viscous dissipation (del2 and del4), and drag and forcing are as follows:
$$
\mathcal{D} = -Ra \, \boldsymbol{u}-C_D \frac{\boldsymbol{u}\left|\boldsymbol{u}\right|}{h}
$$

$$
\mathcal{F} = - C_W \frac{(\boldsymbol{u}_W - \boldsymbol{u})\left|\boldsymbol{u}_W - \boldsymbol{u}\right|}{h}
$$

> The drag consists of simple Rayleigh drag for spin-up as well as quadratic bottom drag. The forcing is a quadratic wind forcing (see equation 1 in [Lilly et al. 2023](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2022MS003327)).

Reviewer 4: the wind forcing seems out of place here.  The form seems fine since it isn't unlike the coupled form, but I'd suggest putting a more generic form like `tau/(rho*h` approach

Author: I don't understand. Dimensionally, the coefficient of a quadratic velocity (normalized by thickness) must be non-dimensional. So the C_D and C_W are non-dimensional. Also, there is no density in a shallow water equation set. 

I was just adding some forcing and dampening terms that would transfer directly to Omega-1. 

FYI, there is a [Stommel basin test case](https://gmd.copernicus.org/articles/16/1297/2023/#section8) with an exact equilibrium solution that includes wind forcing for a shallow water test case. We listed that test as a lower-priority additional test on the [performance testing design document](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-omega0-sw/html/design/OmegaV0ShallowWater.html#inertia-gravity-wave-linearized-shallow-water-no-tracers). I'm sure we could convert that forcing function to whatever form of this coefficient we choose.

Reviewer 4: I guess all I'm saying is that in the barotropic continuous equations the stress would be written as `tau/(rho_w*h)` and what you write in here is an assumed form of tau/rho_w, my suggestion was for the full continuous equation to leave this as `tau/h` and then in the later section specify the quadratic assumption.  Not a big deal to leave as is though.

$$
\left( {\bf k} \cdot \left( \nabla \times {\bf F} \right)\right)_v
\equiv
{\bf k} \cdot \left( \nabla \times F_e \right)
= \frac{1}{A_v} \sum_{e\in EV(v)} t_{e,v} \, F_e \, d_e
$$

where $A_v$ is the area of the dual mesh cell $v$, i.e. the triangle surrounding vertex $v$. Similar to $n_{e,i}$, the indicator function $t_{e,v}$ tracks whether a positive $F_e$ makes a positive or negative contribution to the curl function at vertex $v$. If the vector ${\bf k} \times {\bf n}_e$ is directed toward
${\bf x}_v$, then $t_{e,v}=1$. The summation is over $e\in EV(v)$, which are the edges that terminate at vertex $v$. There are *always* three edges that terminate at each vertex for Voronoi Tessellations, and four edges on each vertex for quadrilateral meshes, unless neighboring cells are missing for land boundaries. Again, the subscript $v$ is dropped and a vertex is assumed for the location of the curl.

#### 3.2.4. Perpendicular vector component

The perpendicular vector component a vector field is defined as
$$
 {\bf F}^\perp \equiv
 {\bf k} \times {\bf F}
$$
so that it points 90$^o$ to the left of ${\bf F}$ in a right-handed coordinate system. 

> In TRiSK, the native component for the prognostic velocity field is the edge-normal velocity $u_e$. The perpendicular component is also positioned at the middle of the edge, but points 90$^o$ to the left, tangential to the edge, from one vertex to another. The perpendicular velocity component is also called the tangential velocity. It is computed diagnostically as

Reviewer 4: is the plan to refer to this velocity as the 'perpendicular' one?  If so we may want to make a comment that in MPAS this is the tangential velocity

Author: I was planning to stay with the `TangentialVelocity` name. Here perpendicular component is just descriptive, but I can clarify that.

Reviewer 4: Ah I see.  Thanks for that clarification

$$
where the ordering of indices $\{i_i, i_2\}$ and $\{v_1, v_2\}$ are explained in the gradient operator sections above.

An alternative formulation for the Del2 dissipation on an unstructured mesh is presented in section 4.2 and Appendix B of 
[Gassman 2018](https://onlinelibrary.wiley.com/doi/10.1002/qj.3294).

#### 3.3.5. Del4 momentum dissipation
The Del4, or biharmonic, momentum dissipation also maps edge-normal velocity back to the edge-normal component of the Laplacian. This is done with two applications of the Del2 operator above.

$$
- \nu_4 \nabla^4 u_e = - \nu_4 \nabla^2 \left( \nabla^2 u_e \right)
$$

#### 3.3.6. Rayleigh Drag
Rayleigh drag is simply linear drag, applied to all levels. It is typically only used during the spin-up process to damp large velocities during the initial adjustment. The Rayleigh coefficient $Ra$ is simply a scalar constant,

$$
- Ra \: u_e.
$$

#### 3.3.7. Bottom drag
Bottom drag is more relevant to layered models than to shallow water systems, but is included here for completeness. It is a quadratic drag applied to the edge velocity,

$$
- C_D \frac{u_e\left|u_e\right|}{[h_i]_e}.
$$

> #### 3.3.8. Wind forcing

Reviewer 4: given this section I would suggest the generic form above and say we choose to model the wind stress as ...

<!--- Note: Table created with [markdown table generator](https://www.tablesgenerator.com/markdown_tables) and original [google sheet](https://docs.google.com/spreadsheets/d/1rz-QXDiwfemq5NpSR1XsvomI7aSKQ1myTNweCY4afcE/edit#gid=0). --->

Table 2. Definition of geometric elements used to build the discrete system.
| Element | Type | Definition | MPAS mesh name | spherical  |
|---------------------|-----------------------------|----------|----------|--------------------------------------------------------------|
| $x_i$ | point | Location of center of primal mesh cells (cell center) | (xCell, yCell, zCell) | (lonCell, latCell) |
| $x_e$ | point | Location of edge points where velocity is defined | (xEdge, yEdge, zEdge) | (lonEdge, latEdge) |
| $x_v$ | point | Location of center of dual-mesh cells (Vertex)| (xVertex, yVertex, zVertex) | (lonVertex, latVertex) |
| $d_e$ | line | segment distance between neighboring $x_i$ locations | dcEdge | |
| $l_e$ | line | segment distance between neighboring $x_v$ locations | dvEdge | |

> | $P_i$ | cell | A cell on the primal mesh | | |
> | $D_v$ | cell | A cell on the dual-mesh | | |

Reviewer 3: ```suggestion
| $P_i$ | cell | A cell on the primal mesh | areaCell | |
| $D_v$ | cell | A cell on the dual-mesh | areaTriangle | |
```

([Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780) eqn 22), 

$$
\left( \nabla h \right)_e
\equiv
\nabla h_i = \frac{1}{d_e} \sum_{i\in CE(e)} -n_{e,i} \, h_i,
$$
where $i\in CE(e)$ identifies the two cells that neighbor an edge. The edge subscript is dropped, so $\nabla h_i$ is understood to occur at an edge.

For practical purposes the sign indicator variable $n_{e,i}$ is not used in the code, and we opt for the simpler formulation
$$
\left( \nabla h \right)_e
\equiv
\nabla h_i = \frac{h_{i2} - h_{i1} }{d_e},
$$
where $\left\{i_1, i_2\right\}$ are the cells neighboring edge $e$.  The indices $\left\{i_1, i_2\right\}$ are ordered such that the normal vector ${\bf n}$ points from cell $i_1$ to cell $i_2$. In the code ${\bf n}$ points from `CellsOnEdge(IEdge, 0)` to `CellsOnEdge(IEdge, 1)`.

#### 3.2.3. Curl
The curl operator maps a vector's edge normal component $u_e$ to a scalar field at vertex
([Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780) eqn 23),

$$
\left( {\bf k} \cdot \left( \nabla \times {\bf F} \right)\right)_v
\equiv
{\bf k} \cdot \left( \nabla \times F_e \right)

> = \frac{1}{A_v} \sum_{e\in EV(v)} t_{e,v} \, F_e \, d_e,

Reviewer 3: ```suggestion
= \frac{1}{{\hat A}_v} \sum_{e\in EV(v)} t_{e,v} \, F_e \, d_e,
```

$$

> where $A_v$ is the area of the dual mesh cell $v$, i.e. the triangle surrounding vertex $v$. Similar to $n_{e,i}$, the indicator function $t_{e,v}$ tracks whether a positive $F_e$ makes a positive or negative contribution to the curl function at vertex $v$. If the vector ${\bf k} \times {\bf n}_e$ is directed toward

Reviewer 3: ```suggestion
where ${\hat A}_v$ is the area of the dual mesh cell $v$, i.e. the triangle surrounding vertex $v$. Similar to $n_{e,i}$, the indicator function $t_{e,v}$ tracks whether a positive $F_e$ makes a positive or negative contribution to the curl function at vertex $v$. If the vector ${\bf k} \times {\bf n}_e$ is directed toward
```

$$

Here ${\bar A}_{e}$ is the rectangle about the edge defined by ${\bar A}_{e} = l_e d_e$ so that
$$
A_i = \sum_{e\in EC(c)} \frac{{\bar A}_{e}}{4}
$$
as stated in 
[Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780) equations 48, 64, and footnote 4.

One might expect an additional factor of one-half in the $K_i$ formula, as in the standard formula for kinetic energy, $K = \tfrac{1}{2} m v^2$. However, the one-half is missing because $u_e$ is only one component of the velocity vector on the edge. One could replace $u_e^2$ with $\tfrac{1}{2}(u_e^2 + {u_e^{\perp}}^2)$, but $u^\perp_e$ is already computed from the surrounding values of $u_e$, so we don't do that because it would just add redundant information. As a side note, the mass $m$ is missing because the mass is already divided out of the momentum equation (1).

The kinetic energy may also be computed at the vertex.  This uses the velocity on the three connected edges,
$$
K_v = \frac{1}{A_v} \sum_{e\in EV(v)} \frac{{\bar A}_{e}}{4}u_e^2.
$$
The normalization can be found through simple geometric arguments to be similar to above,
$$

> A_v = \sum_{e\in EV(v)} \frac{{\bar A}_{e}}{4}.

Reviewer 3: Shouldn't this simply be `{\hat A}_v` ? I don't see this expression in [MPAS-O code](https://github.com/E3SM-Project/E3SM/blob/264a352acdbbdc161db76cfbd3622db8829c07bf/components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F#L1153-L1160). It just uses `areaTriangle`.

Author: Yes. I updated it to `{\hat A}_v`. I just meant to show here the equivalence of the two sides in area. I updated the words to make that more clear.

$$

> The computation of $K_v$ can be found in the MPAS-Ocean code with the `kineticEnergyVertexOnCells` variable. The final kinetic energy at the cell center may be chosen to be either $K_i$ in the first equation above, or a linear combination of the two. In MPAS-Ocean, the flag `config_include_KE_vertex` uses the value

Reviewer 3: ```suggestion
The computation of $K_v$ can be found in the MPAS-Ocean code with the `kineticEnergyVertex` variable. The final kinetic energy at the cell center may be chosen to be either $K_i$ in the first equation above, or a linear combination of `K_i` and `K_v` interpolated to cell centers. In MPAS-Ocean, the flag `config_include_KE_vertex` uses the value
```

$$

> \frac{5}{8} K_i + \frac{3}{8} [K_v]_i

Reviewer 3: Vertex to cell interpolation (`[K_v]_i`) is not defined in the document.

Author: Added that. thanks.

The prognostic velocity variable on the edge is the edge normal velocity, $u_e$. The tangential velocity $u_e^\perp$ is computed diagnostically from $u_e$. In addition, the full vector may be computed at the cell center from the edge normal velocities $u_e$ of the edges on that cell. That is done in MPAS with radial basis functions, and is explained in this [previous design document](https://github.com/MPAS-Dev/MPAS-Documents/blob/master/shared/rbf_design/rbf.pdf).

### 3.3 Momentum Terms

The computation of each term in (4-6) is now described in detail, along with alternative formulations.

#### 3.3.1. Kinetic energy gradient

The kinetic energy gradient term is the non-rotational part of the nonlinear advection. Fundamentally, it maps the prognostic edge-normal velocity from edges back to an edge scalar quantity. We use the standard gradient formulation from cell center to edge,

$$
\nabla K_i = \frac{1}{d_e} \sum_{i\in CE(e)} -n_{e,i}K_i.
$$

There are several methods to compute the cell-centered kinetic energy. The first uses the edge-normal velocities as in
[Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780) equation 63,

$$

> K_i = \frac{1}{A_i} \sum_{e\in EC(c)} \frac{{\bar A}_{e}}{4}u_e^2.

Reviewer 3: ```suggestion
K_i = \frac{1}{A_i} \sum_{e\in EC(i)} \frac{{\bar A}_{e}}{4}u_e^2.
```

$$

Here ${\bar A}_{e}$ is the rectangle about the edge defined by ${\bar A}_{e} = l_e d_e$ so that
$$

> A_i = \sum_{e\in EC(c)} \frac{{\bar A}_{e}}{4}

Reviewer 3: ```suggestion
A_i = \sum_{e\in EC(i)} \frac{{\bar A}_{e}}{4}
```
